### PR TITLE
Error when chunk coordinates are invalid

### LINKED ISSUE #175: Don't allow to write chunks to invalid chunk coordinates

### PR DESCRIPTION
**Pull request by emacollins**
Closes #175

### CODE REVIEW

File: icechunk/src/format/snapshot.rs [hunk 1]
```diff
@@ -49,6 +49,55 @@ pub struct ZarrArrayMetadata {
...
+
+    /// Returns an iterator over the maximum permitted chunk indices for the array.
+    ///
+    /// This function calculates the maximum chunk indices based on the shape of the array
+    /// and the chunk shape.
+    ///
+    /// # Returns
+    ///
+    /// An iterator over the maximum permitted chunk indices.
+    fn max_chunk_indices_permitted(&self) -> impl Iterator<Item = u64> + '_ {
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:62]: I think chunk indices are u32:
```rust
pub struct ChunkIndices(pub Vec<u32>);
```

At the beginning of this method, it could be a good place to:
```rust
debug_assert_eq!(self.shape.len(), self.chunk_shape.0.len())
```
or similar.

File: icechunk/src/format/snapshot.rs [hunk 2]
```diff
@@ -49,6 +49,55 @@ pub struct ZarrArrayMetadata {
...
+        self.shape
+            .iter()
+            .zip(self.chunk_shape.0.iter())
+            .map(|(s, cs)| ((s + cs.get() - 1)) / cs.get() - 1)
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:66]: The math looks a bit weird to me. Shouldn't it be just `s / cs.get() - 1`?

File: icechunk/src/format/snapshot.rs [hunk 3]
```diff
@@ -49,6 +49,55 @@ pub struct ZarrArrayMetadata {
...
+    /// * `coord` - The chunk indices to validate.
+    ///
+    /// # Returns
+    ///
+    /// An `IcechunkResult` indicating whether the chunk coordinates are valid.
+    ///
+    /// # Errors
+    ///
+    /// Returns `IcechunkFormatError::ChunkCoordinatesNotFound` if the chunk coordinates are invalid.
+    pub fn valid_chunk_coord(&self, coord: &ChunkIndices) -> IcechunkResult<bool> {
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:84]: Actually, this function doesn't really fail. Things would be easier if you make it return a simple `bool`

File: icechunk/src/format/snapshot.rs (hunk 1 above)
Comment (emacollins) [reply icechunk/src/format/snapshot.rs]: Pushed a new commit to address this. I ended up returning a `ChunkIndices` type from this using `collect()` on the iterator and type casting the calculation result to u32. Not sure if there is a performance consideration with using collect prior to returning, but it seemed to make sense that the type returned be `ChunkIndices`. This way the actual validation function compares two ChunkIndices.

Shape and chunk shape were both u64, and I assume chunk indices are u32 since they will always be much smaller values. 

Also added the debug assert, which is causing tests in `repository.rs` to fail. I noticed this is because the chunk shape initialized is a single value, even when the shape is multiple. Do we assume that if a single chunk shape value is present, it applies to all dimensions? Or should this fail?

File: icechunk/src/format/snapshot.rs (hunk 2 above)
Comment (emacollins) [reply icechunk/src/format/snapshot.rs]: Yes I don't know why I left this expanded like this. Simplified it down to `(s - 1) / cs.get()`, which should be equivalent and also seems to handle all cases properly. 

I added tests to confirm this, but if there are additional edge cases I haven't thought of, let me know. Will continue to think on it.

File: icechunk/src/format/snapshot.rs [hunk 4]
```diff
@@ -450,4 +502,62 @@ mod tests {
...
+            fill_value: FillValue::Float32(0f32),
+
+            codecs: vec![Codec {
+                name: "mycodec".to_string(),
+                configuration: Some(HashMap::from_iter(iter::once((
+                    "foo".to_string(),
+                    serde_json::Value::from(42),
+                )))),
+            }],
+            storage_transformers: Some(vec![StorageTransformer {
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:527]: I think you can pass `None` here to make it shorter

File: icechunk/src/format/snapshot.rs [hunk 5]
```diff
@@ -450,4 +502,62 @@ mod tests {
...
+                name: "mytransformer".to_string(),
+                configuration: Some(HashMap::from_iter(iter::once((
+                    "foo".to_string(),
+                    serde_json::Value::from(42),
+                )))),
+            }]),
+            dimension_names: Some(vec![
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:534]: same here

File: icechunk/src/format/snapshot.rs [hunk 6]
```diff
@@ -450,4 +502,62 @@ mod tests {
...
+            ]),
+            ..zarr_meta1.clone()
+
+        };
+
+        let coord1 = ChunkIndices(vec![9, 10, 9]);
+        let coord2 = ChunkIndices(vec![10, 11, 10]);
+        let coord3 = ChunkIndices(vec![0, 0 ,0]);
+
+        assert_eq!(true, zarr_meta1.valid_chunk_coord(&coord1));
```
Comment (paraseba) [inline icechunk/src/format/snapshot.rs:557]: replace these by `assert!(...)` or `assert!(! ...)`

File: icechunk/src/repository.rs [hunk 7]
```diff
@@ -424,9 +424,24 @@ impl Repository {
...
+                Err(RepositoryError::FormatError(IcechunkFormatError::ChunkCoordinatesNotFound { coords: coord }))
+            }
+                     
+        } else {
+            Err(RepositoryError::NotAnArray {
+                node: node_snapshot,
+                message: "getting an array".to_string(),
+            })
+        }
+
```
Comment (paraseba) [inline icechunk/src/repository.rs:444]: maybe we can add a test for `Repository` too? that validates that the error is correctly bubbling up.

**Review by paraseba — COMMENTED** [review]
This is looking really great @emacollins

File: icechunk/src/format/snapshot.rs (hunk 1 above)
Comment (paraseba) [reply icechunk/src/format/snapshot.rs]: > I ended up returning a ChunkIndices

perfect!

> I assume chunk indices are u32

yes, you cand valitade looking [here](https://github.com/earth-mover/icechunk/blob/ee982e714d25e1a5681f216141fe95a88f5b2f40/icechunk/src/format/mod.rs#L141)

> chunk shape initialized is a single value, even when the shape is multiple.

this is probably a bad test, we should fix it to match. We need to add more validations to the metadata to avoid this kind of thing, but it's not critical because zarr also validates.

> Do we assume that if a single chunk shape value is present, it applies to all dimensions

No, I don't think so, probably just a bad test, please change it to match.

Comment (emacollins) [reply icechunk/src/format/snapshot.rs]: Thanks for answers! I will return to this PR next week after the holiday.

**Comment by paraseba** [comment]
@emacollins I wonder if the CI failure is a real one ... maybe we are actually writing to an array out of bounds and your change is now flagging it? Do you think you can take a look at what that failing test is doing?

**Comment by paraseba** [comment]
@emacollins I just looked at the code of that test that is failing. I think it's a bad test:
```rust
    let zarr_meta = ZarrArrayMetadata {
        shape: vec![SIZE as u64, SIZE as u64],
        data_type: DataType::Float64,
        chunk_shape: ChunkShape(vec![
            NonZeroU64::new(2).unwrap(),
            NonZeroU64::new(2).unwrap(),
        ]),
        chunk_key_encoding: ChunkKeyEncoding::Slash,
        fill_value: FillValue::Float64(f64::NAN),
        codecs: vec![],
        storage_transformers: None,
        dimension_names: None,
    };

```

Chunks are size 2, and we are writing to all coordinates. I think things will pass if you make the chunk size 1x1

**Comment by paraseba** [comment]
@emacollins please let me know if you need help with that failing test

**Comment by emacollins** [comment]
> @emacollins please let me know if you need help with that failing test

Just pushed the change to run the test again, but caught in some rebase conflicts headache...Will need to pick this up tomorrow. Apologies for the delay on getting this over the line.

**Comment by paraseba** [comment]
No problem @emacollins ! Thank you!

**Review by paraseba — APPROVED** [review]
Fantastic work @emacollins

**Comment by paraseba** [comment]
We just need a `cargo fmt --all` now and we can merge

**Comment by emacollins** [comment]
@paraseba saw some additional tests failed on the last run, pushed a quick fix that should pass them

**Comment by paraseba** [comment]
Awesome work @emacollins, thank you!

Feel free to join our [icechunk slack channel](https://join.slack.com/t/earthmover-community/shared_invite/zt-2cwje92ir-xU3CfdG8BI~4CJOJy~sceQ) if you are interested, and let me know if you want to work on some other issue.
